zng/docs/zson.md

```diff
@@ -0,0 +1,629 @@
+
+### Identifiers
+
+ZSON identifiers are used in several contexts:
+* unquoted field names,
+* names of enum elements, and
+* external type definition names,
+
+Identifiers are case-sensitive and can contain Unicode letters, `$`, `_`,
+and digits (0-9), but may not start with a digit.
+
+### Type Decorators
+
+Any value may be explicitly typed by tagging it with a type decorator.
+The syntax for a decorator is a parenthesized type, as in
+```
+<value> ( <decorator> )
+```
+where a `<decorator>` is either an integer representing an internal type
+or a ZSON type value.  Note that an internal type is not a ZSON type value.
+
+It is an error for the decorator to be type incompatible with its referenced value.  
+
+> TBD: define precisely the notion of type compatibility
+
+#### Type Definitions
+
+Type names are defined by binding a name to a type with an assignment decorator
+of the form
+```
+<value> (= <identifier> )
+```
+This creates a new type whose name is given as the identifier whose value
+is the type value of `<value>`.  This new
+type may then be used as or within a type value or decorator.  The name
+of the type definition
+must not be equal to any of the primitive type names.
+
+The value of a type definition is the given value whose type value is
+given by `<identifier>` as the newly defined type.
+
+It is an error for an external type to be defined to a different type
+than its previously definition though multiple definitions of the same
+type are legal (thereby allowing for concatenation of otherwise
+indepdent sequences).
+
+### Primitive Values
+
+There are 23 types of primitive values with syntax defined as follows:
+
+| Type       | Value Syntax                                                  |
+|------------|---------------------------------------------------------------|
+| `uint8`    | decimal string representation of any unsigned, 8-bit integer  |
+| `uint16`   | decimal string representation of any unsigned, 16-bit integer |
+| `uint32`   | decimal string representation of any unsigned, 32-bit integer |
+| `uint64`   | decimal string representation of any unsigned, 64-bit integer |
+| `int8`     | decimal string representation of any signed, 8-bit integer    |
+| `int16`    | decimal string representation of any signed, 16-bit integer   |
+| `int32`    | decimal string representation of any signed, 32-bit integer   |
+| `int64`    | decimal string representation of any signed, 64-bit integer   |
+| `duration` | a _duration string_ representing signed 64-bit nanoseconds |
+| `time`     | an RFC-3339 UTC data/time string representing signed 64-bit nanoseconds from epoch |
+| `float16`  | a _point string_ representing an IEEE-754 binary16 value |
+| `float32`  | a _point string_ representing an IEEE-754 binary32 value |
+| `float64`  | a _point string_ representing an IEEE-754 binary64 value |
+| `decimal`  | a _point string_ representing an IEEE-754 decimal128 value |
+| `bool`     | the string `true` or `false` |
```

Comment:
FYI, right now the implementation is putting `T` and `F` for these, so I guess one of them needs to change.

```
$ echo '{"foo": true}' | zq -f zson -
{
    foo: T
} (=23)
```